File: 82. Remove Duplicates from Sorted List II/remove-duplicates-from-sorted-list-ii.c
// 82. Remove Duplicates from Sorted List II (4/22/54155)
// Runtime: 7 ms (45.75%) Memory: 6.36 MB (61.64%) 

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* deleteDuplicates (struct ListNode* head) {
    if (!head) {
        return head;
    }
    
    if (head->next && head->next->val == head->val) {
        while (head->next && head->next->val == head->val) {
            head = head->next;
        }
        return deleteDuplicates(head->next);
    }
    else {
        head->next = deleteDuplicates(head->next);
        return head;
    }
}

File: 82. Remove Duplicates from Sorted List II/remove-duplicates-from-sorted-list-ii.ts
// 82. Remove Duplicates from Sorted List II (4/25/54155)
// Runtime: 64 ms (92.39%) Memory: 45.42 MB (11.96%) 

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function deleteDuplicates(head: ListNode | null): ListNode | null {
    if (head === null) {
        return null;
    }
    
    if (head.next && head.next.val === head.val) {
        while (head.next && head.next.val === head.val) {
            head = head.next;
        }
        return deleteDuplicates(head.next);
    } else {
        head.next = deleteDuplicates(head.next);
        return head;
    }
};